README.rst
flask-sillywalk
===============

A Flask extension that implements Swagger support (http://swagger.wordnik.com/)

What's Swagger?
---------------

Swagger is a spec to help you document your APIs. It's flexible and
produces beautiful API documentation that can then be used to build
API-explorer-type sites, much like the one at
http://developer.wordnik.com/docs -- To read more about the Swagger
spec, head over to https://github.com/wordnik/swagger-core/wiki or
http://swagger.wordnik.com

Git Repository and issue tracker: https://github.com/hobbeswalsh/flask-sillywalk

.. |travisci| image::  https://travis-ci.org/hobbeswalsh/flask-sillywalk.png
.. _travisci: https://travis-ci.org/hobbeswalsh/flask-sillywalk

|travisci|_

Why do I want it?
-----------------

* You want your API to be easy to read.
* You want other people to be able to use your API easily.
* You'd like to build a really cool API explorer.
* It's Friday night and your friend just ditched on milkshakes.


How do I get it?
----------------

From your favorit shell:: 

    $ pip install flask-sillywalk


How do I use it?
----------------

I'm glad you asked. In order to use this code, you need to first
instantiate a SwaggerApiRegistry, which will keep track of all your API
endpoints and documentation.

Usage::
    
    from flask import Flask
    from flask.ext.sillywalk import SwaggerApiRegistry, ApiParameter, ApiErrorResponse

    app = Flask("my_api")
    registry = SwaggerApiRegistry(
      app,
      baseurl="http://localhost:5000/api/v1",
      api_version="1.0",
      api_descriptions={"cheese": "Operations with cheese."})
    register = registry.register
    registerModel = registry.registerModel

Then, instead of using the "@app.route" decorator that you're used to
using with Flask, you use the "register" decorator you defined above (or
"registerModel" if you're registering a class that describes a possible
API return value).

Now that we've got an API registry, we can register some functions. The
@register decorator, when just given a path (like @app.route), will
register a GET mthod with no possible parameters. In order to document a
method with parameters, we can feed the @register function some
parameters.

Usage::

    @register("/api/v1/cheese/random")
    def get_random_cheese():
      """Fetch a random Cheese from the database.
      Throws OutOfCheeseException if this is not a cheese shop."""
      return htmlify(db.cheeses.random())

    @register("/api/v1/cheese/<cheeseName>",
      parameters=[
        ApiParameter(
            name="cheeseName",
            description="The name of the cheese to fetch",
            required=True,
            dataType="str",
            paramType="path",
            allowMultiple=False)
      ],
      responseMessages=[
        ApiErrorResponse(400, "Sorry, we're fresh out of that cheese.")
      ])
    def get_cheese(cheeseName):
      """Gets a single cheese from the database."""
      return htmlify(db.cheeses.fetch(name=cheeseName))

Now, if you navigate to http://localhost:5000/api/v1/resources.json you
should see the automatic API documentation. See documentation for all the
cheese endpoints at http://localhost:5000/api/v1/cheese.json


Let's create a view witch uses the ``add_register`` and a model as
``datatype``::

    class Echo(object):
        def __init__(self, name, message="Some message"):
            pass

    def echo():
        data = request.json
        return jsonify(data)

    registry.add_register("/api/v1/echo",
       echo,
       parameters=[
        ApiParameter(
            name="echodata",
            description="Echo me echo",
            required=True,
            dataType="Echo",
            paramType="path",
            allowMultiple=False)
      ],
      method="POST",
      models=[Echo])

What's left to do?
------------------

Well, lots, actually. This release:

* Doesn't support XML (but do we really want to?)
* Doesn't support the full swagger spec (e.g. "type" in data models
* Lots more. Let me know!


CHANGES.rst
=======
Changes
=======

2.1 (unreleased)
----------------
* Add model to register and add_register to get working type.
  [Josip Delic]

* Added add_registerModel.
  [Josip Delic]

* Set content_type to json.
  [Josip Delic]

2.0 (2015-02-12)
----------------

* Python3 support.
  [Josip Delic]

* Added tests, travisci.
  [Josip Delic]

* Proper flask extension.
  [Josip Delic]

* Global registry, but blueprint url rules.
  [Josip Delic]

* Add basepath if not provided.
  [Josip Delic]


flask_sillywalk/tests/test_sillywalk.py
#!/usr/bin/env python
import os
import json
import unittest

from flask import Flask, make_response, request, Blueprint, jsonify
from flask.ext.sillywalk import SwaggerApiRegistry, ApiParameter, ApiErrorResponse
from flask.ext.sillywalk.compat import s


class HappyBirthdayException(Exception):
    pass


class TestDecorators(unittest.TestCase):

    def _create_app(self):
        app = Flask("foobar")
        app.config['TESTING'] = True
        url = os.environ.get("URL", "localhost:5000")
        registry = SwaggerApiRegistry(app, baseurl="http://{0}/api/v1".format(url))

        @app.after_request
        def after_request(data):
            response = make_response(data)
            response.headers['Content-Type'] = 'application/json'
            response.headers['Access-Control-Allow-Origin'] = '*'
            response.headers[
                'Access-Control-Allow-Methods'] = 'POST, GET, OPTIONS , PUT'
            return response

        return app, registry

    def _create_blueprint(self, app):
        bp = Blueprint("blueish", "bluishns")
        app.register_blueprint(bp)
        return bp

    def _create_add_model(self, add_registerModel):
        class SomeCrazyClass(object):
            """This is just the most crazy class!"""

            def __init__(self, name, age, birthday="tomorrow"):
                self.name = name
                self.age = age
                self.birthday = birthday

            def say_happy_birthday(self):
                raise HappyBirthdayException("Chances are it's not your birthday.")

        add_registerModel(SomeCrazyClass)

    def _create_model(self, registerModel):
        @registerModel()
        class SomeCrazyClass(object):
            """This is just the most crazy class!"""

            def __init__(self, name, age, birthday="tomorrow"):
                self.name = name
                self.age = age
                self.birthday = birthday

            def say_happy_birthday(self):
                raise HappyBirthdayException("Chances are it's not your birthday.")

    def _create_add_register(self, add_register, app, bp=None):
        def get_cheese(cheeseName):
            """Gets cheese, just like the name says."""
            return json.dumps({"response": "OK", "message": "Sorry, we're fresh out of {0}!".format(cheeseName)})

        add_register(
            "/api/v1/cheese/<cheeseName>",
            get_cheese,
            parameters=[
                ApiParameter(
                    name="cheeseName",
                    description="The name of the cheese to fetch",
                    required=True,
                    dataType="str",
                    paramType="path",
                    allowMultiple=False)],
            notes='For getting cheese, you know...',
            responseMessages=[
                ApiErrorResponse(400, "Sorry, we're fresh out of that cheese."),
                ApiErrorResponse(418, "I'm actually a teapot")
            ],
            bp=bp)
        # we have register our blueprint again to activate our added route
        if bp:
            app.register_blueprint(bp)

    def _create_add_register_model(self, add_register, app, bp=None):
        class Echo(object):
            def __init__(self, name, message="Some message"):
                pass

        def get_echo():
            data = request.json
            return jsonify(data)

        add_register("/api/v1/echo",
                     get_echo,
                     parameters=[
                         ApiParameter(
                             name="echodata",
                             description="Echo me echo",
                             required=True,
                             dataType="Echo",
                             paramType="path",
                             allowMultiple=False)
                     ],
                     method="POST",
                     models=[Echo])

    def _create_register(self, register, app, bp=None):
        @register(
            "/api/v1/cheese/<cheeseName>",
            parameters=[
                ApiParameter(
                    name="cheeseName",
                    description="The name of the cheese to fetch",
                    required=True,
                    dataType="str",
                    paramType="path",
                    allowMultiple=False)],
            notes='For getting cheese, you know...',
            responseMessages=[
                ApiErrorResponse(400, "Sorry, we're fresh out of that cheese."),
                ApiErrorResponse(418, "I'm actually a teapot")
            ],
            bp=bp)
        def get_cheese(cheeseName):
            """Gets cheese, just like the name says."""
            return json.dumps({"response": "OK", "message": "Sorry, we're fresh out of {0}!".format(cheeseName)})

        # XXX here we are not providing the basepath /api/v1
        @register(
            "/holyHandGrenade/<number>",
            method="GET",
            parameters=[
                ApiParameter(
                    name="number",
                    description="The number of hand grenades to get",
                    required=True,
                    dataType="int",
                    paramType="path",
                    allowMultiple=False)],
            bp=bp)
        def get_a_holy_hand_grenade(number):
            """Gets one or more holy hand grenades, just like the name says."""
            return json.dumps("Fetching {0} holy hand grenades".format(number))

        @register(
            "/api/v1/holyHandGrenade/<number>",
            method="POST",
            parameters=[
                ApiParameter(
                    name="number",
                    description="The number of seconds to wait",
                    required=True,
                    dataType="int",
                    paramType="path",
                    allowMultiple=False),
                ApiParameter(
                    name="target",
                    description="At whom should I thrown the hand grenade?",
                    required=False,
                    dataType="str",
                    paramType="query",
                    allowMultiple=False)],
            bp=bp)
        def toss_the_grenade(number):
            """Toss the holy hand grenade after {number} seconds."""
            target = request.args.get("target", "FOO")
            return json.dumps(
                "Waiting {0} seconds to toss the grenade at {1}.".format(
                    number,
                    target))

        # we have register our blueprint again to activate our added route
        if bp:
            app.register_blueprint(bp)

    def test_model(self):
        app, registry = self._create_app()
        self._create_model(registry.registerModel)
        ret = app.test_client().get("/api/v1/resources.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data['swaggerVersion'], '1.3')
        self.assertEqual(data['basePath'], 'http://localhost:5000/api/v1')
        self.assertEqual(data['apis'], [])
        self.assertEqual(data['apiVersion'], '1.0')
        self.assertEqual(data['models']['SomeCrazyClass']['properties'],
                         {u'age': {u'required': True, u'type': u'string'},
                          u'birthday': {u'defaultValue': u'tomorrow', u'type': u'string'},
                          u'name': {u'required': True, u'type': u'string'}})

    def test_add_model(self):
        app, registry = self._create_app()
        self._create_add_model(registry.add_registerModel)
        ret = app.test_client().get("/api/v1/resources.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data['swaggerVersion'], '1.3')
        self.assertEqual(data['basePath'], 'http://localhost:5000/api/v1')
        self.assertEqual(data['apis'], [])
        self.assertEqual(data['apiVersion'], '1.0')
        self.assertEqual(data['models']['SomeCrazyClass']['properties'],
                         {u'age': {u'required': True, u'type': u'string'},
                          u'birthday': {u'defaultValue': u'tomorrow', u'type': u'string'},
                          u'name': {u'required': True, u'type': u'string'}})

    def test_register(self):
        app, registry = self._create_app()
        self._create_register(registry.register, app)
        ret = app.test_client().get("/api/v1/resources.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data['swaggerVersion'], '1.3')
        self.assertEqual(data['basePath'], 'http://localhost:5000/api/v1')
        self.assertEqual(sorted([x["path"] for x in data['apis']]),
                         ['/cheese.{format}', '/holyHandGrenade.{format}'])
        self.assertEqual(data['apiVersion'], '1.0')
        self.assertEqual(data['models'], {})
        # test cheese
        ret = app.test_client().get("/api/v1/cheese.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data["resourcePath"], "cheese")
        self.assertEqual([x.endpoint for x in app.url_map.iter_rules()],
                         ['holyHandGrenade',
                          'resources',
                          'cheese',
                          'get_a_holy_hand_grenade',
                          'toss_the_grenade',
                          'get_cheese',
                          'static'])
        self.assertEqual([x.rule for x in app.url_map.iter_rules()],
                         ['/api/v1/holyHandGrenade.json',
                          '/api/v1/resources.json', '/api/v1/cheese.json',
                          '/api/v1/holyHandGrenade/<number>',
                          '/api/v1/holyHandGrenade/<number>',
                          '/api/v1/cheese/<cheeseName>',
                          '/static/<path:filename>'])

    def test_add_register(self):
        app, registry = self._create_app()
        self._create_add_register(registry.add_register, app)
        ret = app.test_client().get("/api/v1/resources.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data['swaggerVersion'], '1.3')
        self.assertEqual(data['basePath'], 'http://localhost:5000/api/v1')
        self.assertEqual(sorted([x["path"] for x in data['apis']]),
                         ['/cheese.{format}'])
        self.assertEqual(data['apiVersion'], '1.0')
        self.assertEqual(data['models'], {})
        # test cheese
        ret = app.test_client().get("/api/v1/cheese.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data["resourcePath"], "cheese")

    def test_add_register_model(self):
        app, registry = self._create_app()
        self._create_add_register_model(registry.add_register, app)
        ret = app.test_client().get("/api/v1/resources.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data['swaggerVersion'], '1.3')
        self.assertEqual(data['basePath'], 'http://localhost:5000/api/v1')
        self.assertEqual(sorted([x["path"] for x in data['apis']]),
                         ['/echo.{format}'])
        self.assertEqual(data['apiVersion'], '1.0')
        self.assertEqual(data['models'], {})
        # test cheese
        ret = app.test_client().get("/api/v1/echo.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data["resourcePath"], "echo")
        self.assertEqual(data["models"].keys(), ["Echo"])

    def test_register_blueprint(self):
        app, registry = self._create_app()
        bp = self._create_blueprint(app)
        self._create_register(registry.register, app, bp)
        ret = app.test_client().get("/api/v1/resources.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data['swaggerVersion'], '1.3')
        self.assertEqual(data['basePath'], 'http://localhost:5000/api/v1')
        self.assertEqual(sorted([x["path"] for x in data['apis']]),
                         ['/cheese.{format}', '/holyHandGrenade.{format}'])
        self.assertEqual(data['apiVersion'], '1.0')
        self.assertEqual(data['models'], {})
        # test cheese
        ret = app.test_client().get("/api/v1/cheese.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data["resourcePath"], "cheese")
        # FIXME this is broken urls shouldn't be added multiple times
        self.assertEqual([x.endpoint for x in app.url_map.iter_rules()],
                         ['holyHandGrenade',
                          'resources',
                          'cheese',
                          'blueish.get_a_holy_hand_grenade',
                          'blueish.toss_the_grenade',
                          'blueish.get_cheese',
                          'static'])

    def test_add_register_blueprint(self):
        app, registry = self._create_app()
        bp = self._create_blueprint(app)
        self._create_add_register(registry.add_register, app, bp)
        ret = app.test_client().get("/api/v1/resources.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data['swaggerVersion'], '1.3')
        self.assertEqual(data['basePath'], 'http://localhost:5000/api/v1')
        self.assertEqual(sorted([x["path"] for x in data['apis']]),
                         ['/cheese.{format}'])
        self.assertEqual(data['apiVersion'], '1.0')
        self.assertEqual(data['models'], {})
        # test cheese
        ret = app.test_client().get("/api/v1/cheese.json")
        data = json.loads(s(ret.data))
        self.assertEqual(data["resourcePath"], "cheese")
        self.assertEqual([x.endpoint for x in app.url_map.iter_rules()],
                         ['resources',
                          'cheese',
                          'blueish.get_cheese',
                          'static'])


flask_sillywalk/sillywalk.py
import inspect

from collections import defaultdict
from flask.ext.sillywalk.compat import urlparse
from flask import jsonify


__SWAGGERVERSION__ = "1.3"
SUPPORTED_FORMATS = ["json"]


class SwaggerRegistryError(Exception):
    """A Swagger registry error"""
    pass


class SwaggerApiRegistry(object):
    """
    If you're going to make your Python/Flask API swagger compliant, you'll need
    to initialize a SwaggerApiRegistry with a Flask app and a baseurl.

    >>> my_app = Flask(__name__)
    >>> registry = SwaggerApiRegistry(my_app, "http://my_url.com/api/v1")

    Then you can register URLs with this class' "register" method.
    """

    def __init__(self, app=None, baseurl="http://localhost/",
                 api_version="1.0", api_descriptions={}):
        self.baseurl = baseurl
        self.api_version = api_version
        self.api_descriptions = api_descriptions
        self.basepath = urlparse(self.baseurl).path
        self.r = defaultdict(dict)
        self.models = defaultdict(dict)
        self.registered_routes = []
        if app is not None:
            self.app = app
            self.init_app(self.app)

    def init_app(self, app):
        """
        Initialize the Flask app by adding the base "resources" URL. Currently only JSON
        is supported, so this will add the URL <baseurl>/resources.json to your app.
        """
        for fmt in SUPPORTED_FORMATS:
            app.add_url_rule(
                "{0}/resources.{1}".format(
                    self.basepath.rstrip("/"),
                    fmt),
                "resources",
                self.jsonify(self.resources))

    def jsonify(self, f):
        """
        In case we need to serialize different stuff in the future.
        """

        def inner_func():
            return jsonify(f())

        return inner_func

    def resources(self):
        """
        Gets all currently known API resources and serialized them.
        """
        resources = {
            "apiVersion": self.api_version,
            "swaggerVersion": __SWAGGERVERSION__,
            "basePath": self.baseurl,
            "models": dict(),
            "apis": list()}
        for resource in self.r.keys():
            description = (self.api_descriptions[resource]
                           if resource in self.api_descriptions else "")
            resources["apis"].append({
                "path": "/" + resource + ".{format}",
                "description": description})
        for k, v in self.models.items():
            resources["models"][k] = v
        return resources

    def _class_to_model(self, c, type_="object"):
        if self.app is None:
            raise SwaggerRegistryError(
                "You need to initialize {0} with a Flask app".format(
                    self.__class__.__name__))
        model = {
            "id": c.__name__,
            "description": c.__doc__ if c.__doc__ is not None else "",
            "type": type_,
            "properties": dict()}
        argspec = inspect.getargspec(c.__init__)
        argspec.args.remove("self")
        defaults = {}
        nondefaults = argspec.args

        if argspec.defaults:
            defaults = list(zip(argspec.args[-len(
                argspec.defaults):], argspec.defaults))
        if defaults:
            nondefaults = argspec.args[:-len(defaults)]

        for arg in nondefaults:
            model["properties"][arg] = {
                "required": True, "type": "string"}
        for k, v in defaults:
            model["properties"][k] = {
                "defaultValue": v, "type": "string"}
        return model

    def _registerModel(self, c, type_):
        model = self._class_to_model(c, type_)
        self.models[c.__name__] = model
        return c

    def add_registerModel(self,
                          c,
                          type_="object"):
        """
        Registers a Swagger Model (object).

        By default every argument is set as ``string``.

        TODO

        add __fields__ to specify more paramaters::

              __fields__ = {
                  'name': {'defaultValue': 'Hans',
                           'required': True,
                           'type': 'number'},
              }

        Usage:

        >>> class Dog(object):
        >>>     def __init__(self, name, status="ok"):
        >>>         pass
        >>> my_registry.add_registerModel(Dog, type="Animal")
        """
        self._registerModel(c, type_)

    def registerModel(self,
                      type_="object"):
        """
        Registers a Swagger Model (object).

        By default every argument is set as ``string``.

        TODO

        add __fields__ to specify more paramaters::

              __fields__ = {
                  'name': {'defaultValue': 'Hans',
                           'required': True,
                           'type': 'number'},
              }

        Usage:

        >>> @my_registry.registerModel(type="Animal")
        >>> class Dog(object):
        >>>     def __init__(self, name, status="ok"):
        >>>         pass
        """

        def inner_func(c):
            self._registerModel(c, type_)
        return inner_func

    def _register(self,
                  path,
                  f,
                  method,
                  content_type,
                  parameters,
                  responseMessages,
                  nickname,
                  notes,
                  bp,
                  models):

        if self.app is None:
            raise SwaggerRegistryError(
                "You need to initialize {0} with a Flask app".format(
                    self.__class__.__name__))

        # use basepath if not set by user
        if self.basepath not in path:
            path = self.basepath + "/" + path
            path = path.replace("//", "/")

        # register views on blueprints
        app = self.app

        if bp:
            app = bp

        app.add_url_rule(
            path,
            f.__name__,
            f,
            methods=[method])

        api = Api(
            method=f,
            path=path.replace(self.basepath, ""),
            httpMethod=method,
            params=parameters,
            responseMessages=responseMessages,
            nickname=nickname,
            notes=notes,
            models=models)

        if api.resource not in self.app.view_functions:
            for fmt in SUPPORTED_FORMATS:
                route = "{0}/{1}.{2}".format(self.basepath.rstrip("/"),
                                             api.resource, fmt)
                if route not in self.registered_routes:
                    self.registered_routes.append(route)
                    self.app.add_url_rule(
                        route,
                        api.resource,
                        self.jsonify(self.show_resource(api.resource)))

        if self.r[api.resource].get(api.path) is None:
            self.r[api.resource][api.path] = list()
        self.r[api.resource][api.path].append(api)

    def add_register(self,
                     path,
                     f,
                     method="GET",
                     content_type="application/json",
                     parameters=[],
                     responseMessages=[],
                     nickname=None,
                     notes=None,
                     bp=None,
                     models=[]):
        """
        Registers an API endpoint.

        Usage:

        >>> def get_cheese(cheesename):
        >>>     # some function
        >>> my_registry.add_register(
        ...     "/api/v1/cheese/<cheeseName>",
        ...     get_cheese,
        ...     parameters=[ApiParameter(
        ...         name="cheeseName",
        ...         description="The name of the cheese to fetch",
        ...         required=True,
        ...         dataType="str",
        ...         paramType="path",
        ...         allowMultiple=False)],
        ...     notes='For getting cheese, you know...',
        ...     responseMessages=[
        ...         ApiErrorResponse(400, "Sorry, we're fresh out of that cheese."),
        ...         ApiErrorResponse(418, "I'm actually a teapot")]))

        """
        self._register(path, f, method, content_type, parameters,
                       responseMessages, nickname, notes, bp, models)

    def register(self,
                 path,
                 method="GET",
                 content_type="application/json",
                 parameters=[],
                 responseMessages=[],
                 nickname=None,
                 notes=None,
                 bp=None,
                 models=[]):
        """
        Registers an API endpoint.

        Usage:

        >>> @my_registry.register(
        ...     "/api/v1/cheese/<cheeseName>",
        ...     parameters=[ApiParameter(
        ...         name="cheeseName",
        ...         description="The name of the cheese to fetch",
        ...         required=True,
        ...         dataType="str",
        ...         paramType="path",
        ...         allowMultiple=False)],
        ...     notes='For getting cheese, you know...',
        ...     responseMessages=[
        ...         ApiErrorResponse(400, "Sorry, we're fresh out of that cheese."),
        ...         ApiErrorResponse(418, "I'm actually a teapot")]))
        >>> def get_cheese(cheesename):
        >>>     # some function

        """
        def inner_func(f):
            self._register(path, f, method, content_type, parameters,
                           responseMessages, nickname, notes, bp, models)
        return inner_func

    def show_resource(self, resource):
        """
        Serialize a single resource.
        """

        def inner_func():
            return_value = {
                "resourcePath": resource.rstrip("/"),
                "apiVersion": self.api_version,
                "swaggerVersion": __SWAGGERVERSION__,
                "basePath": self.baseurl,
                "apis": list(),
                "models": dict()
            }
            resource_map = self.r.get(resource)
            for path, apis in resource_map.items():
                api_object = {
                    "path": path,
                    "description": "",
                    "operations": list()}
                for api in apis:
                    for m in api.models:
                        return_value["models"][m.__name__] = self._class_to_model(m)
                    api_object["operations"].append(api.document())
                return_value["apis"].append(api_object)
            return return_value

        return inner_func


class SwaggerDocumentable(object):
    """
    A documentable swagger object, e.g. an API endpoint, an API parameter, an API error response...
    """

    def document(self):
        return self.__dict__


class Api(SwaggerDocumentable):
    """
    A single API endpoint.
    """

    def __init__(
            self,
            method,
            path,
            httpMethod,
            params=None,
            responseMessages=None,
            nickname=None,
            notes=None,
            models={}):
        self.httpMethod = httpMethod
        self.summary = method.__doc__ if method.__doc__ is not None else ""
        self.resource = path.lstrip("/").split("/")[0]
        self.path = path.replace("<", "{").replace(">", "}")
        self.parameters = [] if params is None else params
        self.responseMessages = [] if responseMessages is None else responseMessages
        self.nickname = "" if nickname is None else nickname
        self.notes = notes
        self.models = models

    # See https://github.com/wordnik/swagger-core/wiki/API-Declaration
    def document(self):
        ret = self.__dict__.copy()
        # need to serialize these guys
        ret.pop("models")
        ret["parameters"] = [p.document() for p in self.parameters]
        ret["responseMessages"] = [e.document() for e in self.responseMessages]
        return ret

    def __hash__(self):
        return hash(self.path)


class ApiParameter(SwaggerDocumentable):
    """
    A single API parameter
    """

    def __init__(
            self,
            name,
            description,
            required,
            dataType,
            paramType,
            allowMultiple=False):
        self.name = name
        self.description = description
        self.required = required
        self.dataType = dataType
        self.paramType = paramType
        self.allowMultiple = allowMultiple

    def document(self):
        return self.__dict__


class ImplicitApiParameter(ApiParameter):
    """
    Not sure what I was thinking here... --hobbeswalsh
    """

    def __init__(self, *args, **kwargs):
        if "default_value" not in kwargs:
            raise TypeError(
                "You need to provide an implicit param with a default value.")
        super(ImplicitApiParameter, self).__init__(*args, **kwargs)
        self.defaultValue = kwargs.get("default_value")


class ApiErrorResponse(SwaggerDocumentable):
    """
    An API error response.
    """

    def __init__(self, code, message):
        self.message = message
        self.code = code


example.py
#!/usr/bin/env python
import os
import json

from flask import Flask, make_response, request, jsonify
from flask.ext.sillywalk import SwaggerApiRegistry, ApiParameter, ApiErrorResponse


app = Flask("cheeseShop")

url = os.environ.get("URL", "localhost:5000")
registry = SwaggerApiRegistry(app, baseurl="http://{0}/api/v1".format(url))
register = registry.register
registerModel = registry.registerModel


class HappyBirthdayException(Exception):
    pass


@registerModel()
class SomeCrazyClass(object):
    """This is just the most crazy class!"""

    def __init__(self, name, age, birthday="tomorrow"):
        self.name = name
        self.age = age
        self.birthday = birthday

    def say_happy_birthday(self):
        raise HappyBirthdayException("Chances are it's not your birthday.")


@register(
    "/api/v1/cheese/<cheeseName>",
    parameters=[
        ApiParameter(
            name="cheeseName",
            description="The name of the cheese to fetch",
            required=True,
            dataType="str",
            paramType="path",
            allowMultiple=False)],
    notes='For getting cheese, you know...',
    responseMessages=[
        ApiErrorResponse(400, "Sorry, we're fresh out of that cheese."),
        ApiErrorResponse(418, "I'm actually a teapot")
    ])
def get_cheese(cheeseName):
    """Gets cheese, just like the name says."""
    return json.dumps(
        {"response": "OK", "message": "Sorry, we're fresh out of {0}!".format(
            cheeseName)
        })


@register(
    "/api/v1/holyHandGrenade/<number>",
    method="GET",
    parameters=[
        ApiParameter(
            name="number",
            description="The number of hand grenades to get",
            required=True,
            dataType="int",
            paramType="path",
            allowMultiple=False)])
def get_a_holy_hand_grenade(number):
    """Gets one or more holy hand grenades, just like the name says."""
    return json.dumps("Fetching {0} holy hand grenades".format(number))


@register(
    "/api/v1/holyHandGrenade/<number>",
    method="POST",
    parameters=[
        ApiParameter(
            name="number",
            description="The number of seconds to wait",
            required=True,
            dataType="int",
            paramType="path",
            allowMultiple=False),
        ApiParameter(
            name="target",
            description="At whom should I thrown the hand grenade?",
            required=False,
            dataType="str",
            paramType="query",
            allowMultiple=False)])
def toss_the_grenade(number):
    """Toss the holy hand grenade after {number} seconds."""
    target = request.args.get("target", "FOO")
    return json.dumps(
        "Waiting {0} seconds to toss the grenade at {1}.".format(
            number,
            target))


class Echo(object):
    def __init__(self, name, message="Some message"):
        pass


def get_echo():
    data = request.json
    return jsonify(data)


registry.add_register("/api/v1/echo",
                      get_echo,
                      parameters=[
                          ApiParameter(
                              name="echodata",
                              description="Echo me echo",
                              required=True,
                              dataType="Echo",
                              paramType="path",
                              allowMultiple=False)
                      ],
                      method="POST",
                      models=[Echo])


@app.after_request
def after_request(data):
    response = make_response(data)
    response.headers['Content-Type'] = 'application/json'
    response.headers['Access-Control-Allow-Origin'] = '*'
    response.headers[
        'Access-Control-Allow-Methods'] = 'POST, GET, OPTIONS , PUT'
    return response


port = int(os.environ.get("PORT", 5000))
# print(app.url_map)
app.run(host="0.0.0.0", port=port, debug=True)
